libs/core/type_support/include/hpx/type_support/relocate_at.hpp

```diff
@@ -0,0 +1,101 @@
+//  Copyright (c) 2023 Isidoros Tsaousis-Seiras
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/type_support/is_trivially_relocatable.hpp>
+#include <hpx/type_support/construct_at.hpp>
+
+#include <cstring>
+#include <type_traits>
+
+namespace hpx {
+
+#if defined(HPX_HAVE_P1144_STD_RELOCATE_AT)
+    using std::relocate_at;    // what would be the include for this?
+    using std::relocate;
+#else
+
+    namespace detail {
+        template <class T>
+        struct destroy_guard
+        {
+            T& t;
+            explicit destroy_guard(T& t)
+              : t(t)
+            {
+            }
+            ~destroy_guard()
+            {
+                t.~T();
```

Comment:
`destroy_guard` should just take `T*` (not `T&`), for simplicity.
It should arguably call `hpx::destroy_at(t)` here, just in case `T` is an array type (see below).
I'm making both of these changes to my libc++ code as we speak.